Added more helper functions to `ImplicitTileCoordinates`

IanLilleyT opened the pull request:
While reviewing https://github.com/CesiumGS/cesium/pull/9517 I started to integrate the implicit tiling code into another feature I'm working on. The transition went really smoothly except I wanted more capability out of `ImplicitTileCoordinates`. So this PR is mainly about that, as well as some little fixes I found along the way.

cesium-concierge commented:
Thanks for the pull request @IanLilleyT!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

IanLilleyT commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -926,7 +926,7 @@ function Cesium3DTileset(options) {
    * a {@link Cesium3DTilesetMetadata} object to access metadata.
    *
-   * @type {Cesium3DTilesetMetadata}
+   * @type {Cesium3DTilesetMetadata|undefined}
```
The metadata can be undefined, so I marked it

IanLilleyT commented on Source/Scene/ImplicitSubtree.js:
```diff
@@ -81,7 +81,7 @@ Object.defineProperties(ImplicitSubtree.prototype, {
 
   /**
-   * When the 3DTILES_metadata extension is used, this property stores
+   * When the <code>3DTILES_metadata</code> extension is used, this property stores
```
A small documentation inconsistency

IanLilleyT commented on Source/Scene/ImplicitSubtree.js:
```diff
@@ -108,19 +108,46 @@ Object.defineProperties(ImplicitSubtree.prototype, {
+   * @readonly
+   * @private
+   */
+  implicitCoordinates: {
```
This was useful to have for the more manual implicit tile coordinate processing that I'm doing

IanLilleyT commented on Source/Scene/ImplicitSubtree.js:
```diff
@@ -108,19 +108,46 @@ Object.defineProperties(ImplicitSubtree.prototype, {
+ * @returns {Boolean} the value of the i-th bit
+ * @private
+ */
+ImplicitSubtree.prototype.tileIsAvailableAtCoordinates = function (
```
Wanted the ability to query by coordinates as well. This way external systems don't need to know what morton encoding is

ptrgags replied:
Yeah these additions/changes to the interface make sense now that we have more requirements for implicit tiling.

IanLilleyT commented on Source/Scene/ImplicitSubtree.js:
```diff
@@ -108,19 +108,46 @@ Object.defineProperties(ImplicitSubtree.prototype, {
  * @private
  */
-ImplicitSubtree.prototype.tileIsAvailable = function (index) {
+ImplicitSubtree.prototype.tileIsAvailableAtIndex = function (index) {
```
Changes the name to differentiate from the `Coordinates` version

IanLilleyT commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -121,7 +161,7 @@ Object.defineProperties(ImplicitTileCoordinates.prototype, {
 
   /**
-   * Get the Morton index for this tile within the current LOD by interleaving
+   * Get the Morton index for this tile within the current level by interleaving
```
naming consistency

IanLilleyT commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -25,7 +25,8 @@ import ImplicitSubdivisionScheme from "./ImplicitSubdivisionScheme.js";
  * @param {Object} options An object with the following properties:
  * @param {ImplicitSubdivisionScheme} options.subdivisionScheme Whether the coordinates are for a quadtree or octree
- * @param {Number} options.level The level of a tile relative to the tile with the extension.
+ * @param {Number} options.subtreeLevels The number of distinct levels within the coordinate's subtree
```
Adding `subtreeLevels` made a lot of the new helper functions possible. This has been updated everywhere that uses `ImplicitTileCoordinates`

IanLilleyT commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -178,12 +219,322 @@ ImplicitTileCoordinates.prototype.deriveChildCoordinates = function (
+ * @returns {Boolean} <code>true</code> if this tile is an ancestor of the other tile
+ * @private
+ */
+ImplicitTileCoordinates.prototype.isAncestorOf = function (otherCoordinates) {
```
Open for a better name if there is one

IanLilleyT commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -178,12 +219,322 @@ ImplicitTileCoordinates.prototype.deriveChildCoordinates = function (
+ * @returns {Boolean} <code>true</code> if this tile is the root
+ * @private
+ */
+ImplicitTileCoordinates.prototype.isRoot = function () {
```
Is this too ambiguous? There's `isRootOfSubtree` but the difference could be easy to miss

IanLilleyT commented on Source/Scene/ImplicitTileset.js:
```diff
@@ -16,13 +16,17 @@ import ImplicitSubdivisionScheme from "./ImplicitSubdivisionScheme.js";
  * @experimental This feature is using part of the 3D Tiles spec that is not final and is subject to change without Cesium's standard deprecation policy.
  */
-export default function ImplicitTileset(tileset, baseResource, tileJson) {
+export default function ImplicitTileset(
```
Removing the dependency on a `Cesium3DTileset` frees other systems to use this class. All uses of `ImplicitTileset` have been updated to match the new constructor.

IanLilleyT commented on Specs/Scene/ImplicitSubtreeSpec.js:
```diff
@@ -30,7 +29,9 @@ describe("Scene/ImplicitSubtree", function () {
     var expectedAvailability = availabilityToBooleanArray(availability);
     for (var i = 0; i < availability.lengthBits; i++) {
-      expect(subtree.tileIsAvailable(i)).toEqual(expectedAvailability[i]);
+      expect(subtree.tileIsAvailableAtIndex(i)).toEqual(
```
This file is missing coverage for `tileIsAvailableAtCoordinates` and `childSubtreeIsAvailableAtCoordinates`. This could be a good place to put them, but it was annoying to derive the coordinate from `i`. I wonder if there's a better way to do coverage for those new functions.

ptrgags replied:
would it help to add some static functions to `ImplicitTilingTester` if you find yourself making the same coordinate calculation over and over?

ptrgags replied:
or just a helper function in this file if this is the only place that uses it.

IanLilleyT commented on Specs/Scene/ImplicitTileCoordinatesSpec.js:
```diff
@@ -209,7 +1501,8 @@ describe("Scene/ImplicitTileCoordinates", function () {
       subdivisionScheme: ImplicitSubdivisionScheme.OCTREE,
-      level: 4,
+      subtreeLevels: 8,
+      level: 6,
```
This test used to be invalid because it was level 4 and had a z value of 32. Had to raise the level to fit.

ptrgags commented on Specs/Scene/ImplicitTilesetSpec.js:
```diff
@@ -38,13 +38,13 @@ describe("Scene/ImplicitTileset", function () {
       baseResource,
-      implicitTileJson
+      implicitTileJson,
+      metadataSchema
```
Also, it would be good to add a test that this metadata is accessible if this does get passed in. I've been putting things in different describe blocks by extension, so that would probably go best in something like:

```
describe("3DTILES_metadata", function() {
  it("stores metadataSchema", function() {
     // construct an implicit tileset and check for metadataSchema
  });
}):
```

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -178,12 +219,322 @@ ImplicitTileCoordinates.prototype.deriveChildCoordinates = function (
+  var tileX = this.x;
+  var tileY = this.y;
+
+  var subtreeLevel = ((tileLevel / subtreeLevels) | 0) * subtreeLevels;
```
Wait `NaN | 0` results in `0`? This is news to me. Please put a comment explaining this to avoid confusion.

ptrgags replied:
Also wait a sec is the intent of this something along the lines of `Math.floor(tileLevel / subtreeLevels) * subtreeLevels`? Definitely explain this more thoroughly, it took me a while.

IanLilleyT replied:
I'll just go with `Math.floor`

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -178,12 +219,322 @@ ImplicitTileCoordinates.prototype.deriveChildCoordinates = function (
+};
+
+/**
+ * Get the coordinates of the subtree that contains this tile
```
Could we make sure that in the documentation of these functions that we're clear what are local and what are global coordinates?

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -178,12 +219,322 @@ ImplicitTileCoordinates.prototype.deriveChildCoordinates = function (
+ * @returns {ImplicitTileCoordinates} The local coordinates of the child subtree
+ * @private
+ */
+ImplicitTileCoordinates.prototype.deriveLocalChildSubtreeCoordinates = function () {
```
This method name feels too long, though I'm also struggling to find a concise way to say "subtree root coordinates relative to its parents"

ptrgags replied:
Also another thing I notice about this file is we have a lot of functions thrown around, but the symmetries are a bit unclear.

We might want to discuss this in more depth, maybe over a whiteboard to plan the most concise set of orthogonal functions here.

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -178,12 +219,322 @@ ImplicitTileCoordinates.prototype.deriveChildCoordinates = function (
+};
+
+/**
+ * Returns whether this tile is at the bottom of the subtree
```
Maybe mention that this is the last row of nodes, but _not_ the child subtrees.

ptrgags commented on Source/Scene/ImplicitSubtree.js:
```diff
@@ -696,19 +737,39 @@ function makeJumpBuffer(subtree) {
 /**
  * Get the entity ID for a tile within this subtree.
- * @property {Cesium3DTile} tile A transcoded tile from this subtree. tile.implicitTileCoordinates must be defined.
+ * @property {ImplicitTileCoordinates} implicitCoordinates The coordinates of a tile
```
Another place where we should be explicit about local/global.

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -25,7 +25,8 @@ import ImplicitSubdivisionScheme from "./ImplicitSubdivisionScheme.js";
  *
```
@IanLilleyT and I had a whiteboard session to design a cleaner interface for this file. 

![image](https://user-images.githubusercontent.com/8422414/116926586-85e55280-ac28-11eb-8791-8ab35df10977.png)

The diagram was a bit of a mess by the end, so here are the key functions we need:

**Fundamental functions**:

All three of these functions work whether using global coordinates (relative to the root implicit tile) or local (relative to a subtree root)

* `ImplicitTileCoordinates.prototype.deriveDescendant(offsetCoordinates)` - computes the coordinates of a tile deeper in the tree by combining the current coordinates + an offset. It's roughly analogous to vector addition `position + offset` going _down_ the tree.
* `ImplicitTileCoordinates.prototype.deriveAnscestor(offsetLevels)` - computes the coordinates of the tile higher up in the tree by going up a number of levels. Since this is on a tree, there's only ever one possible path so only a level offset is needed. This is also analogous to vector addition `position + offset` except going _up_ the tree.
* `ImplicitTileCoordinates.getRelativeCoordinates(anscestor, descendant)` (not sure the best signature, or if it should be a static method or an instance method) - compute the coordinates of descendant relative to ancestor. This is analogous to vector subtraction `descendant - anscestor`

**Other Functions**:

* `ImplicitTileCoordinates.prototype.deriveSubtreeCoordinates()` - exists for convenience, built atop `deriveAnscestor` using `offsetLevels = tileLevel - subtreeLevel`.
* `ImplicitTileCoordinates.prototype.deriveParentSubtreeCoordinates()` - exists for convenience, built atop `deriveAncestor`. `offsetLevels = (tileLevel - subtreeLevel) + subtreeLevels`
* `ImplicitTileCoordinates.deriveChildCoordinates(childIndex)` - We still need this since this is optimized for the transcoder (document it as such)
* `deriveLocalChildSubtreeCoordinates()` - actually we didn't say if we still needed this, but you can build it on top of `getRelativeCoordinates()` in any case. Though if we do keep this function, see if we can find a shorter name...

IanLilleyT replied:
`deriveLocalChildSubtreeCoordinates` could be implemented as `deriveRelativeCoordinates(tile.deriveParentSubtreeCoordinates(), tile)` and `deriveLocalTileCoordinates` could be implemented as `deriveRelativeCoordinates(tile.deriveSubtreeCoordinates(), tile)`

If that's too much to expect out of someone using the API, or too error-prone, we could have a helper function called `deriveChildSubtreeCoordinates` and `deriveLocalTileCoordinates`?

IanLilleyT replied:
Since you mention static vs instance, all function that take two inputs should be consistent. I think the API will be cleaner if it's all instance methods and it's made clear what is what. What do you think?

ptrgags replied:
For the first point, how often will those functions be used?

Hm, as far as `deriveRelativeCoordinates`, I guess the usage would normally be `descendant.deriveRelativeCoordinates(anscestor)`? or is the other way around better?

IanLilleyT replied:
hmm, the only place it's used is in `ImplicitSubtree`
```
ImplicitSubtree.prototype.childSubtreeIsAvailableAtCoordinates = function (
  implicitCoordinates
) {
  var index = this.getChildSubtreeIndex(implicitCoordinates);
  return this.childSubtreeIsAvailableAtIndex(index);
};
```
could become
```
ImplicitSubtree.prototype.childSubtreeIsAvailableAtCoordinates = function (
  implicitCoordinates
) {
  var subtreeCoordinates = implicitCoordinates.deriveParentSubtreeCoordinates();
  var offsetCoordinates = subtreeCoordinates.deriveOffset(implicitCoordinates);
  var index = this.getChildSubtreeIndex(offsetCoordinates);
  return this.childSubtreeIsAvailableAtIndex(index);
};
```

IanLilleyT commented on Source/Scene/ImplicitSubtree.js:
```diff
@@ -695,20 +736,30 @@ function makeJumpBuffer(subtree) {
+ * @return {Number} The child subtree's index within the subtree's child subtree availability bitstream.
+ * @private
+ */
+ImplicitSubtree.prototype.getChildSubtreeIndex = function (
```
:laughing: :+1:

IanLilleyT commented:
@ptrgags I believe I've addressed all of the feedback, both big and small. 

The API for `ImplicitTileCoordinates` is now:

`getDescendantCoordinates(offsetCoordinates)`
`getAncestorCoordinates(offsetLevel)`
`getOffsetCoordinates(descendantCoordinates)`
`getChildCoordinates(mortonIndex)` - a special case of `getDescendantCoordinates`
`getSubtreeCoordinates()` - a special case of `getAncestorCoordinates`
`getParentSubtreeCoordinates()` - a special case of `getAncestorCoordinates`
`isAncestor(descendantCoordinates)`
`isEqual(otherCoordinates)`
`isImplicitTilesetRoot()`
`isSubtreeRoot()`
`isBottomOfSubtree()`
`fromMortonIndex(level, mortonIndex)` / `.mortonIndex`
`fromTileIndex(tileIndex)` / `.tileIndex`
`.childIndex` - equivalent to `getAncestorCoordinates(1).getOffsetCoordinates(this).mortonIndex`

I called it `offset` instead of `relative` because `relative` is perhaps too familial. But I'm ok with changing the name to something else for this or any of the other functions.

Also, there are now unit tests and coverage for all new functions in `ImplicitTileCoordinates` and `ImplicitSubtree`

ptrgags commented on Specs/Scene/ImplicitTileCoordinatesSpec.js:
```diff
@@ -39,24 +61,347 @@ describe("Scene/ImplicitTileCoordinates", function () {
+
+    // mismatched subtreeLevels
+    expect(function () {
+      return new ImplicitTileCoordinates({
+        subdivisionScheme: ImplicitSubdivisionScheme.QUADTREE,
+        subtreeLevels: 2,
+        level: 0,
+        x: 0,
+        y: 0,
```
Is there a reason why `internalQuadtree()` isn't used here like in many of the tests above? I do like the concise helper function

ptrgags commented on Specs/Scene/ImplicitTileCoordinatesSpec.js:
```diff
@@ -4,9 +4,30 @@ import {
 } from "../../Source/Cesium.js";
 
 describe("Scene/ImplicitTileCoordinates", function () {
+  function internalQuadtree(level, x, y) {
```
why the name `internal`? I thought you meant internal node, but I've seen you use it with (0, 0, 0).

IanLilleyT replied:
I was mimicking after `internalQuadtreeDescription` in `ImplicitSubtreeSpec`. I used to have it be `buildQuadtree` so maybe I'll set it back to that to avoid confusion.

ptrgags replied:
ohhh that "internal" meant subtree with internal buffers (rather than external buffers referenced by uri). not relevant here.

maybe `quadtreeCoordinates`? `buildQuadree` sounds like you're making a whole tree of nodes

ptrgags commented on Specs/Scene/ImplicitTileCoordinatesSpec.js:
```diff
@@ -4,9 +4,30 @@ import {
+  function internalQuadtree(level, x, y) {
+    return new ImplicitTileCoordinates({
+      subdivisionScheme: ImplicitSubdivisionScheme.QUADTREE,
+      subtreeLevels: 2,
```
could subtreeLevels be an optional argument at the end? that way you could use it everywhere in this fil.

IanLilleyT replied:
Good idea. I'll use the helper function everywhere except for `constructs quadtree` and `constructs octree`

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -35,12 +37,42 @@ import ImplicitSubdivisionScheme from "./ImplicitSubdivisionScheme.js";
 export default function ImplicitTileCoordinates(options) {
```
Hm, maybe the class documentation should include a paragraph explaining what we mean by global (relative to the implicit root tile) and local (relative to the root of a specific subtree) coordinates?

ptrgags replied:
Also maybe include some notes about the limitations on the range level, x, y, z imposed by the bit twiddling and `MortonOrder`?

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -136,20 +177,194 @@ Object.defineProperties(ImplicitTileCoordinates.prototype, {
+ * @param {ImplicitTileCoordinates} b
+ * @private
+ */
+function checkMatchingProperties(a, b) {
```
this name sounds more general than it is. Maybe `checkMatchingSubtreeShape(a, b)`?

The way I think of it is the subdivision scheme is like the "width" of the subtree and subtreeLevels is the "height"

I feel like we're going to get a theory of tree geometry by the time we merge this PR :laughing:

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -178,12 +394,128 @@ ImplicitTileCoordinates.prototype.deriveChildCoordinates = function (
+    return false;
+  }
+
+  var isAncestorX = tileX === descendantCoordinates.x >> levelDifference;
+  var isAncestorY = tileY === descendantCoordinates.y >> levelDifference;
```
Minor nitpick, I feel like this would be a little easier to understand if we didn't combine the shift and the `===` in one line, what do you think?

```suggestion
  var ancestorX = descendantCoordinates.x >> levelDifference;
  var ancestorY = descendantCoordinates.y >> levelDifference;
  ...
  return tileX == ancestorX && tileY == ancestorY;
```

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -204,18 +536,21 @@ ImplicitTileCoordinates.prototype.getTemplateValues = function () {
  * octree/quadtree, compute the (level, x, y, [z]) coordinates
  *
  * @param {ImplicitSubdivisionScheme} subdivisionScheme
+ * @param {Number} subtreeLevels
```
missing description

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -237,8 +573,60 @@ ImplicitTileCoordinates.fromMortonIndex = function (
+ * the (level, x, y, [z]) coordinates
+ *
+ * @param {ImplicitSubdivisionScheme} subdivisionScheme
+ * @param {Number} subtreeLevels
```
same here.

ptrgags commented on Source/Scene/ImplicitTileCoordinates.js:
```diff
@@ -237,8 +573,60 @@ ImplicitTileCoordinates.fromMortonIndex = function (
+  var mortonIndex;
+
+  if (subdivisionScheme === ImplicitSubdivisionScheme.OCTREE) {
+    // Node count up to octree level: (8^L - 1) / (8-1)
+    // (8^L - 1) / (8-1) <= X < (8^(L+1) - 1) / (8-1)
+    // 8^L <= (7x + 1) < 8^(L+1)
+    // L <= log8(7x + 1) < L + 1
+    // L = floor(log8(7x + 1))
+    // L = floor(log2(7x + 1) / log2(8))
+    // L = floor(log2(7x + 1) / 3)
```
I'm happy to see these derivations in a comment :+1:

ptrgags reviewed (commented):
@IanLilleyT Looks a lot cleaner than before! I still have some feedback, but it's mostly minor details

IanLilleyT commented:
@ptrgags this is ready for another review

ptrgags commented on Specs/Scene/ImplicitTilesetSpec.js:
```diff
@@ -38,14 +39,15 @@ describe("Scene/ImplicitTileset", function () {
+      implicitTileJson,
+      metadataSchema
     );
+    expect(implicitTileset.metadata).toBeUndefined();
```
```suggestion
    expect(implicitTileset.metadataSchema).toBeUndefined();
```

ptrgags reviewed (approved):
@IanLilleyT looks solid. I noticed one small typo, but I'll push a commit for that and merge once CI passes.

Thanks!

ptrgags commented:
Travis is being difficult today. The only issue was a build timed out, but Travis is having issues with the GH API so I can't log in to restart it

ptrgags commented:
After talking with @lilleyse, we're just going to merge this into the staging branch.
